UefiPayloadPkg/Library/PciSegmentInfoLibAcpiBoardInfo/PciSegmentInfoLibAcpiBoardInfo.c

```diff
@@ -13,48 +15,220 @@
+  //
+  PciRootBridgeInfo = (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES *)GET_GUID_HOB_DATA (GuidHob);
+  if (PciRootBridgeInfo->Count <= (GET_GUID_HOB_DATA_SIZE (GuidHob) - sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES)) / sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGE)) {
+    return PciRootBridgeInfo;
+  }
+
+  return NULL;
+}
+
+/**
+  Return info for all root bridge segments
+
+  @return All the segment info instances in an array.
+**/
+UPL_PCI_SEGMENT_INFO_HOB *
+Get_UPLSegInfo (
+  VOID
+  )
+{
+  UPL_PCI_SEGMENT_INFO_HOB          *UplSegmentInfo;
+  EFI_HOB_GUID_TYPE                 *GuidHob;
+  UNIVERSAL_PAYLOAD_GENERIC_HEADER  *GenericHeader;
+
+  //
+  // Find Universal Payload Segment Info hob
+  //
+  GuidHob = GetFirstGuidHob (&gUplPciSegmentInfoHobGuid);
+  if ((GuidHob == NULL) || (sizeof (UNIVERSAL_PAYLOAD_GENERIC_HEADER) > GET_GUID_HOB_DATA_SIZE (GuidHob))) {
+    return NULL;
+  }
+
+  GenericHeader = (UNIVERSAL_PAYLOAD_GENERIC_HEADER *)GET_GUID_HOB_DATA (GuidHob);
+  if (GenericHeader->Length > GET_GUID_HOB_DATA_SIZE (GuidHob)) {
+    return NULL;
+  }
+
+  if ((GenericHeader->Revision != UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES_REVISION) || (GenericHeader->Length < sizeof (UPL_PCI_SEGMENT_INFO_HOB))) {
+    return NULL;
+  }
+
+  //
+  // UPL_PCI_SEGMENT_INFO_HOB structure is used when Revision equals to UPL_PCI_SEGMENT_INFO_HOB_REVISION
+  //
+  UplSegmentInfo = (UPL_PCI_SEGMENT_INFO_HOB *)GET_GUID_HOB_DATA (GuidHob);
+  if (UplSegmentInfo->Count <= (GET_GUID_HOB_DATA_SIZE (GuidHob) - sizeof (UPL_PCI_SEGMENT_INFO_HOB)) / sizeof (UPL_SEGMENT_INFO)) {
+    return UplSegmentInfo;
+  }
+
+  return NULL;
+}
+
+/**
+  Return all the root bridge instances in an array.
+
+  @param Count  Return the count of root bridge instances.
+
+  @return All the root bridge instances in an array.
+          The array should be passed into PciHostBridgeFreeRootBridges()
+          when it's not used.
 **/
 PCI_SEGMENT_INFO *
 EFIAPI
 GetPciSegmentInfo (
   UINTN  *Count
   )
 {
-  EFI_HOB_GUID_TYPE  *GuidHob;
-  ACPI_BOARD_INFO    *AcpiBoardInfo;
+  UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES  *PciRootBridgeInfo;
+  UPL_PCI_SEGMENT_INFO_HOB            *UplSegmentInfo;
 
-  ASSERT (Count != NULL);
-  if (Count == NULL) {
-    return NULL;
+  if (mPciSegments != NULL) {
+    *Count = mCount;
+    return mPciSegments;
   }
 
-  if (mPciSegment0.BaseAddress == 0) {
-    //
-    // Find the acpi board information guid hob
-    //
-    GuidHob = GetFirstGuidHob (&gUefiAcpiBoardInfoGuid);
-    ASSERT (GuidHob != NULL);
+  PciRootBridgeInfo = Get_RBInfo ();
+  UplSegmentInfo    = Get_UPLSegInfo ();
+  if (PciRootBridgeInfo == NULL) {
+    return 0;
+  }
 
-    AcpiBoardInfo            = (ACPI_BOARD_INFO *)GET_GUID_HOB_DATA (GuidHob);
-    mPciSegment0.BaseAddress = AcpiBoardInfo->PcieBaseAddress;
+  if (UplSegmentInfo == NULL) {
+    RetrieveSegmentInfoFromHob (PciRootBridgeInfo, Count);

```

Comment:
Thanks for supporting backward compatibility!
We will verify this new implementation, meantime, do we really need to add hard dependency of PciRootBridgeInfo to old model function RetrieveSegmentInfoFromHob() when it does not consume at all? I think we may move "PciRootBridgeInfo = Get_RBInfo ();" to else path below as only new model function consuming it. What do you think?

Reply:
done